services/edgecluster/k3s/k3s.go

```diff
@@ -0,0 +1,294 @@
+		service.logger.Fatal("failed to validate the requested namespace",
+			zap.Any("Error", err))
+		return err
+	}
+
+	if ns != nil && ns.GetName() == namespace {
+		service.logger.Info("the namespace exists")
+		return nil
+	}
+
+	return nil
+}
+
+func getRestConfig(logger *zap.Logger) (*rest.Config, error) {
+	kubeconfig := os.Getenv("KUBECONFIG")
+
+	logger.Info("path ",
+		zap.String("KUBECONFIG", kubeconfig))
+
+	if kubeconfig != "" {
+		return clientcmd.BuildConfigFromFlags("", kubeconfig)
+	}
+
+	homePath, err := os.UserHomeDir()
+
+	logger.Info("homePath ",
+		zap.String("path", homePath))
+	if err != nil {
+		return nil, err
+	}
+
+	kubeconfigFilePath := filepath.Join(homePath, ".kube", "config")
+
+	logger.Info("kubePath ",
+		zap.String("kube path", kubeconfigFilePath))
+
+	_, err = os.Stat(kubeconfigFilePath)
+	if !os.IsNotExist(err) {
+		return clientcmd.BuildConfigFromFlags("", kubeconfigFilePath)
+	}
+
+	return rest.InClusterConfig()
+}
+
+func makeDeploymentConfig(request *types.NewProvisionRequest) *appsv1.Deployment {
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      request.Name + "-deployment",
+			Namespace: request.NameSpace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: int32Ptr(DeploymentReplica),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": request.Name,
+				},
+			},
+			Template: apiv1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": request.Name,
+					},
+				},
+				Spec: apiv1.PodSpec{
+					Containers: []apiv1.Container{
+						{
+							Name:  ContainerName,
+							Image: ContainerImage,
+							Args: []string{
+								"server",
+								"--disable-agent",
+								"--advertise-address=" + request.ContainerIpAddress,
+							},
+							Ports: []apiv1.ContainerPort{
+								{
+									Name:          "http",
+									Protocol:      apiv1.ProtocolTCP,
+									ContainerPort: DeploymentContainerPort,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return deployment
+}
+
+func makeServiceConfig(request *types.NewProvisionRequest) *apiv1.Service {
+
+	servicePorts := []v1.ServicePort{
+		{
+			Protocol:   apiv1.ProtocolTCP,
+			Port:       request.ServicePort,
+			TargetPort: intstr.FromInt(request.TargetPort),
+		},
+	}
+
+	serviceSelector := map[string]string{
+		"app": request.Name,
+	}
+
+	service := &apiv1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      request.Name + "-service",
+			Namespace: request.NameSpace,
+			Labels: map[string]string{
+				"k8s-app": request.Name,
+			},
+		},
+		Spec: apiv1.ServiceSpec{
+			Ports:     servicePorts,
+			Selector:  serviceSelector,
+			ClusterIP: "",
+		},
+	}
+
+	return service
+}
+
+func makeNameSpaceConfig(namespace string) *v1.Namespace {
+	ns := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespace,
+		},
+	}
+
+	return ns
+}
+
+func int32Ptr(i int32) *int32 {
```

Comment:
Is this really necessary? Can't we just use the variable address in the only place where the function is actually called?